src/librustc/ty/sty.rs

```diff
@@ -2151,64 +2151,60 @@ impl<'tcx> Const<'tcx> {
     }
 
     #[inline]
-    pub fn to_bits(
-        &self,
-        tcx: TyCtxt<'_, '_, 'tcx>,
-        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
-    ) -> Option<u128> {
-        if self.ty != ty.value {
-            return None;
-        }
-        let ty = tcx.lift_to_global(&ty).unwrap();
-        let size = tcx.layout_of(ty).ok()?.size;
-        self.val.try_to_bits(size)
-    }
-
-    #[inline]
-    pub fn to_ptr(&self) -> Option<Pointer> {
-        self.val.try_to_ptr()
-    }
-
-    #[inline]
-    pub fn assert_bits(
+    pub fn try_eval_bits(
         &self,
         tcx: TyCtxt<'_, '_, '_>,
-        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
+        ty: Ty<'_>,
     ) -> Option<u128> {
-        assert_eq!(self.ty, ty.value);
+        assert_eq!(self.ty, ty);
+        let tcx = tcx.global_tcx();
         let ty = tcx.lift_to_global(&ty).unwrap();
-        let size = tcx.layout_of(ty).ok()?.size;
-        self.val.try_to_bits(size)
+        // FIXME(eddyb, oli-obk) get the right param_env.
```

Comment:
Can you keep the hacks at callsites? Most calls should not need to use `ParamEnv::empty()`.

Reply:
Looking at this, at least everything `InferCtxt` does has a `ParamEnv`:
https://github.com/rust-lang/rust/blob/3f36ac4e835dfa0af77d5261a444a674ec9da3a5/src/librustc/infer/combine.rs#L44-L48
Here's an example of where `TypeRelation` could have a method to get it:
https://github.com/rust-lang/rust/blob/3f36ac4e835dfa0af77d5261a444a674ec9da3a5/src/librustc/infer/equate.rs#L25-L30

cc @nikomatsakis @varkor Can we let `TypeRelation` know the `ParamEnv`?

Reply:
I made `TypeRelation` have a method that returns the `ParamEnv` (although I left a few cases as `ParamEnv::empty()` and added a FIXME.

Now I'm stuck with trying to lift the following `ParamEnv` to the global tcx:

```
error: internal compiler error: src/librustc/ty/sty.rs:2163: cannot lift ParamEnvAnd {                                                                                                                       
    param_env: ParamEnv {
        caller_bounds: [
            Binder(
                OutlivesPredicate(
                    '_#0r,
                    '_#1r
                )
            )
        ],
        reveal: UserFacing,
        def_id: None
    },
    value: usize
} for Const {
    ty: usize,
    val: Scalar(
        Bits {
            size: 8,
            bits: 2
        }
    )
}
```

In this case the `ParamEnv` is pretty useless considering that all we're doing is to try to compute the layout of `usize`. We can't lift it because it only exists in the local interner as it was created on the fly for

```
#0 [evaluate_obligation] evaluating trait selection obligation `&[&str; 2]: marker::Copy`
```

Reply:
You need to use canonicalization in the `const_eval` query to do this then - which is the correct approach. It means inference variables are replaced with a kind of "type parameter".
Since `const_eval` doesn't actually do any inference itself, you could treat `ty::Placeholder` (I think?) as `ty::Param`, but I'm not sure, @nikomatsakis should comment on this.

You could search for `canonicalize_query` around to find some examples, like this:
https://github.com/rust-lang/rust/blob/54479c624cc24bb852291d83e11cc19389dcb21b/src/librustc/traits/query/outlives_bounds.rs#L99-L130

Reply:
As for your failure... you could work around it for now by having a non-ICE error path for failing to lift the `param_env, just like for failing to lift the `substs`, e.g.:
https://github.com/rust-lang/rust/blob/54479c624cc24bb852291d83e11cc19389dcb21b/src/librustc/traits/select.rs#L781

and (but this tries them separately for a reason not applicable in your case):

https://github.com/rust-lang/rust/blob/54479c624cc24bb852291d83e11cc19389dcb21b/src/librustc/traits/fulfill.rs#L462-L467

Reply:
I considered that, but I don't think it's going to go well. We have perfectly evaluable constants, that are actually evaluated today (due to `ParamEnv::empty()`, and suddenly they would stop being evaluable even though they are "obviously" fine. If we suddenly can't check for zero length arrays, a bunch of type system things will change (e.g. `[String; 0]` would stop being promotable).

I can try it out, but I can't really judge the consequences of doing so.

Reply:
Ahhh, you should be doing something like this then:
https://github.com/rust-lang/rust/blob/54479c624cc24bb852291d83e11cc19389dcb21b/src/librustc/traits/project.rs#L404-L406

That is, falling back to trying to evaluate the constant with identity substitutions. You can also get `tcx.param_env(def_id)` for the "identity `ParamEnv`" (which is the most general, too; both the substs and the param env should be empty today, anyway, for things like array lengths).

(I'm not sure why I wasn't more general in the linked code, heh)